File: 3sum.swift
#!/usr/bin/env swift
// http://programmingpraxis.com/2013/06/18/3sum/

func findZeroSumTerms(array: [Int]) -> [[Int]] {
  var solutions:[String:[Int]] = [:]
  for i in 0..<array.count {
    for j in i..<array.count {
      let testVal = -(array[i] + array[j])
      if array.contains(testVal) {
        let threeToAppend = [array[i], array[j], testVal].sort { $0 < $1 }
        solutions["\(threeToAppend)"] = threeToAppend
      }
    }
  }
  return Array(solutions.values)
}

let testArray = [8, -25, 4, 10, -10, -7, 2, -3] 
print(findZeroSumTerms(testArray))

File: 3_sum.py
#!/usr/bin/env python
# http://programmingpraxis.com/2013/06/18/3sum/

def find_zero_sum_terms(list):
  solutions = []
  list_length = len(list)
  for outer_index in xrange(0,list_length-1):
    for inner_index in xrange(outer_index,list_length-1):
      test_value = -(list[outer_index] + list[inner_index])
      if test_value in list:
        solutions.append([list[outer_index],list[inner_index],test_value])
  sorted_solutions = []
  for triplet in solutions:
    t = sorted(triplet)
    if t in sorted_solutions:
      continue
    else:
      sorted_solutions.append(t)
  return sorted_solutions

test_array = [8, -25, 4, 10, -10, -7, 2, -3]

print find_zero_sum_terms(test_array)
